kale/pipeline/mida_trainer.py

```diff
@@ -0,0 +1,201 @@
+        transformer (BaseEstimator, optional): Transformer object implementing 'fit_transform'. Defaults to None.
+        mida_param_grid (dict, optional): Grids for searching the optimal hyperparameters of MIDA. Defaults to None.
+        transformer_param_grid (dict, optional): Grids for searching the optimal hyperparameters of the transformer.
+            Defaults to None.
+    """
+
+    def __init__(
+        self,
+        estimator,
+        estimator_param_grid=None,
+        transformer=None,
+        mida_param_grid=None,
+        transformer_param_grid=None,
+    ):
+        self.estimator = estimator
+        self.transformer = transformer
+        self.mida_param_grid = mida_param_grid
+        self.transformer_param_grid = transformer_param_grid
+        self.estimator_param_grid = estimator_param_grid
+        self.best_transformer = None
+        self.best_transformer_params = None
+        self.best_mida = None
+        self.best_mida_params = None
+        self.best_estimator = None
+        self.best_estimator_params = None
+        self.best_score = None
+
+    def fit(self, x, y, groups=None):
+        """Fit a Transformer -> Maximum independence domain adaptation -> Estimator pipeline with the given data x ,
+            labels y, and group labels groups
+
+        Args:
+            x (array-like): Training input data matrix, shape (n_samples, n_features)
+            y (array-like): Labels, shape (n_labeled_samples,)
+            groups (array-like, optional): Domain/group labels of x, shape (n_samples,)
+
+        Returns:
+            self
+        """
+        if self.transformer is None:
+            self.transformer = FunctionTransformer()
+        if self.transformer_param_grid is None:
+            self.transformer_param_grid = {key: [val] for key, val in self.transformer.get_params().items()}
+        transformer_grid = ParameterGrid(self.transformer_param_grid)
+        for transformer_params in transformer_grid:
+            self.transformer.set_params(**transformer_params)
+            x_transformed = self.transformer.fit_transform(x)
+            self._fit_mida(x_transformed, y, groups, transformer_params)
+
+        self.best_transformer = self.transformer.set_params(**self.best_transformer_params)
+        self.best_mida = MIDA(**self.best_mida_params)
+        self.best_estimator = self.estimator.set_params(**self.best_estimator_params)
+
+        x_transformed = self.best_transformer.transform(x)
+        x_transformed = self.best_mida.fit_transform(x_transformed, y=y, groups=groups)
+        self.best_estimator.fit(x_transformed[y.shape[0] :,], y)
+
+        return self
+
+    def _fit_mida(self, x_transformed, y, groups, transformer_params):
+        """Fit MIDA and estimator with the given data x_transformed and labels
+
+        Args:
+            x_transformed (array-like): Transformed data, shape (n_samples, n_features)
+            y (array-like): Data labels, shape (n_labeled_samples,)
```

Comment:
I think just allowing unlabeled `y` to be -1 is a preferable option. we could easily index the labeled data when the label and unlabeled data is in a random order.

Reply:
it will make our assumptions easier when doing `_validate_data` as it expects `X` and `y` to have the same number of samples.